File: leetcode/linked-lists/143-reorder-list/c/solution.c
#include <stdlib.h>

/// {{
/// Problem: 143. Reorder List
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/reorder-list/
/// Topics: `linked-list`, `two-pointers`, `stack`, `recursion`
/// Timestamp: `Wed, 25 Jun 2025 10:42:34 +0530`

// Definition for singly-linked list.
struct ListNode {
  int val;
  struct ListNode *next;
};

void reorderList(struct ListNode *head) {
  if (head == NULL || head->next == NULL || head->next->next == NULL) {
    return;
  }

  struct ListNode *prev, *fast, *slow;

  // Get to the middle node
  for (
    fast = head, prev = NULL, slow = head;
    fast != NULL && fast->next != NULL;
    fast = fast->next->next, prev = slow, slow = slow->next
  ) { }

  // Break the list at the middle node
  prev->next = NULL;

  struct ListNode *cur, *next;

  // Reverse the second sub-list
  for (
    prev = NULL, cur = slow, next = NULL;
    cur != NULL;
    cur = next
  ) {
    next = cur->next;
    cur->next = prev;
    prev = cur;
  }

  struct ListNode dest = {.val = 0, .next = NULL};
  struct ListNode *p1, *p2, *p3;

  for (
    p1 = head, p2 = prev, p3 = &dest;
    p1 != NULL || p2 != NULL;
  ) {
    p3->next = p1;
    if (p3->next != NULL) {
      p3 = p3->next;
    }
    if (p1 != NULL) {
      p1 = p1->next;
    }

    p3->next = p2;
    if (p3->next != NULL) {
      p3 = p3->next;
    }
    if (p2 != NULL) {
      p2 = p2->next;
    }
  }

  if (head->next != NULL) {
    head->next = dest.next->next;
  }
}

/// }}

File: leetcode/linked-lists/143-reorder-list/ts/solution.ts
/// {{
/// Problem: 143. Reorder List
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/reorder-list/
/// Timestamp: `Thu, 18 Sep 2025 20:33:13 +0530`

// Definition for singly-linked list.
class ListNode {
  val: number
  next: ListNode | null
  constructor(val?: number, next?: ListNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }
}

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
  if (!head) {
    return;
  }

  const middle = middleNode(head);
  if (!middle) {
    return;
  }

  const next = middle.next;
  middle.next = null;

  const head2 = reverseList(next);
  let [n1, n2]: (ListNode)[] = [head, head2];

  while (n1 && n2) {
    const next1 = n1.next;
    const next2 = n2.next;

    n1.next = n2;
    n2.next = next1;
    n1 = next1;
    n2 = next2;
  }
}

function middleNode(head: ListNode | null): ListNode | null {
  if (!head) {
    return null;
  }
  let [slow, fast]: (ListNode | null)[] = [head, head];
  for (; fast && fast.next; fast = fast.next.next) {
    slow = slow?.next ?? null;
  }
  return slow;
}

function reverseList(head: ListNode | null): ListNode | null {
  if (!head) {
    return null;
  }
  let [prev, cur, next]: (ListNode | null)[] = [null, head, head.next];
  for (; cur; prev = cur, cur = next) {
    next = cur.next;
    cur.next = prev;
  }
  return prev;
}

/// }}

